x-pack/filebeat/input/awss3/s3_filters.go

```diff
@@ -0,0 +1,124 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package awss3
+
+import (
+	"sync"
+	"time"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+const (
+	filterOldestTime = "oldestTimeFilter"
+	filterStartTime  = "startTimeFilter"
+)
+
+// filterProvider exposes filters that needs to be applied on derived state.
+// Once configured, retrieve filter applier using getApplierFunc.
+type filterProvider struct {
+	cfg *config
+
+	staticFilters []filter
+	once          sync.Once
+}
+
+func newFilterProvider(cfg *config) *filterProvider {
+	fp := &filterProvider{
+		cfg: cfg,
+	}
+
+	// derive static filters
+	if cfg.StartTimestamp != "" {
+		// note - errors should not occur as this has validated prior reaching here
+		parse, _ := time.Parse(time.RFC3339, cfg.StartTimestamp)
+		fp.staticFilters = append(fp.staticFilters, newStartTimestampFilter(parse))
+	}
+
+	return fp
+}
+
+// getApplierFunc returns aggregated filters valid at the time of retrival.
+// Applier return true if state is valid for processing according to the underlying filter collection.
+func (f *filterProvider) getApplierFunc() func(log *logp.Logger, s state) bool {
+	filters := map[string]filter{}
```

Comment:
Can we cache this instead of initializing it in every run?

We could initialize it in the `new` function, and add the `filterStartTime` if needed.  Then in this function we could remove it, and this way we could get rid of the `sync.Once` too.

Reply:
I think we can't. I had to define two types of filters, static and dynamic. The static ones we can configure once and use repeatedly. The dynamic ones require re-initialization when retrieving. `oldestTimeFilter` is one such (implementing `ignore_older`) where we need to consider `time.Now` to filter out inputs. 

And, I had to rely on `sync.Once` to avoid `oldestTimeFilter` at first retrieval only so that we can allow processing entries from `start_timestamp` when it is defined.

Reply:
I think the way to have it static would require a change like:

```go
func newOldestTimeFilter(ignoreOlder time.Duration) *oldestTimeFilter {
	return &oldestTimeFilter{
		id:         filterOldestTime,
		ignoreOlder: ignoreOlder,
	}
}

func (s oldestTimeFilter) isValid(objState state, time time.Time) bool {
	return <time - objState time within ignore older> 
}
```

But that's your choice, of course. I'm fine with either.

Reply:
Thanks for the suggestion. Yes it can work. But I think I will keep the current implementation so that in future if we see the need for a new filter we can add it easily :)